### DISCUSSION TITLE
Multiple flet apps in a nginx proxy_pass subfolder structure - is that at all possible?

### DISCUSSION THREAD

**Discussion by rkleivel** [flet-dev/flet#1760] [General]
### /
So - I made my awesome dashboard with flet and published it in a docker container called *dash*. App is started with:
`ft.app(name="", port=8850, target=main, view=ft.WEB_BROWSER)`

My nginx reverse proxy (in another docker container) is configured like described in the [Self Hosting Guide](https://flet.dev/docs/guides/python/deploying-web-app/hosting-providers/self-hosting) (only adapted `proxy_pass` to my environment), and calling https://app.mydomain.com/ works like charm. The dashboard works as expected. Nginx config looks like this:
```
    location / {
      proxy_pass http://dash:8850/;
      proxy_http_version 1.1;
      proxy_set_header   Upgrade $http_upgrade;
      proxy_set_header   Connection keep-alive;
      proxy_set_header   Host $host;
      proxy_cache_bypass $http_upgrade;
      proxy_set_header   X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header   X-Forwarded-Proto $scheme;
    }


    location /ws {
      proxy_pass http://dash:8850/ws;
      proxy_http_version 1.1;
      proxy_set_header   Upgrade $http_upgrade;
      proxy_set_header   Connection "upgrade";
      proxy_set_header   Host $host;
      proxy_cache_bypass $http_upgrade;
      proxy_set_header   X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header   X-Forwarded-Proto $scheme;
    }
```
### /dash
Next step is that **I don't really want my dashboard under in the root folder `/` as I plan to have multiple flet dashboards in `/dash`, `/dash2` ..., and maybe a toolbox in plain html in the root folder `/` with links to the dashboards**
![Screenshot from 2023-08-24 14-09-28](https://github.com/flet-dev/flet/assets/15140187/d2470786-f497-47bf-a852-238c00eb769c)

To move my first dashboard from `/` to `/dash` I found that the nginx config needs to look like this to work (shortened to show changes or important areas only):

```
    # https://serverfault.com/questions/1101045/nginx-serve-application-under-sub-path
    if ($http_referer ~ ^https?://app.mydomain.com/dash) {
        rewrite ^ /dash:8850$uri;
    }

    location /dash {
      proxy_pass http://dash:8850/;
      .
      .
    }

    location /ws {
      proxy_pass http://dash:8850/ws;
      .
      .
    }
```
With this config my dashboard works fine on https://app.mydomain.com/dash. The changes I made are **1)** adding the rewrite block and **2)** change the app location from `/` to `/dash` SO FAR SO GOOD

BUT WAIT - why didn't I change the location for websocket from `/ws` to `/dash/ws` as you would expect just following the pattern? Because that did not work. And I did try adding another rewrite block like this:
```
 #   if ($http_referer ~ ^https?://app2.dev.bergenengines.com/dash/ws) {
 #       rewrite ^ /dash:8850/ws$uri;
 #   }
```
### /dash and /dash2
I am by no means any expert on nginx nor websocket. But testing confirms that I cannot add the second dashboard by just adding a second rewrite and location like this

```
    if ($http_referer ~ ^https?://app.mydomain.com/dash2) {
        rewrite ^ /dash2:8850$uri;
    }

    location /dash2 {
      proxy_pass http://dash2:8850/;
      .
      .
    }
```
because I would also need to route websocket to the 2 different containers dash and dash2, and that is not possible as long as the websocket locations would be `/ws` for both of them.

### So to the question(s):
* Can someone help me achieve my goal?
* Maybe it is possible to customize the websocket endpoints paths in flet? (I found [this commit](https://github.com/flet-dev/flet/pull/1756/commits/9af7feda782cbe788204cb2f9f2510815089e97d) that I *think* suggests it should be possible)
* Maybe some other nginx config could manage to distinguish the traffic to each of the websockets?
* Any hint appreciated :)

**Comment by FeodorFitsner** [comment]
This is exactly [the feature I'm working on](https://github.com/flet-dev/flet/pull/1756) right now: hosting multiple Flet web apps with production-grade durability!
In this scenario FastAPI is replacing built-in Fletd web server and the entire setup is run by gunicorn/uvicorn. WebSocket connections are managed by Starlette/FastAPI and Flet program "lives" inside it. One of the requirements to a Flet program though - it must be async in order to work with FastAPI.

I'm doing final touches in that PR and hopefully merge it tomorrow, so we all will have a chance to give it a try.
To get prepared start adjusting your app(s) to be async: https://flet.dev/docs/guides/python/async-apps

It's exciting to see Flet handling dozens of sessions/connections while gunicorn has only one thread running! :)

**Reply by rkleivel** [reply]
Ah  - that's great news 😊👍 Looking fwd to that feature!
